File: openi40.scheduler/angular-app/projects/openi40-scheduler-ui/src/lib/timeutils/TimeSegment.ts
/**
 * 
 * This code is part of the OpenI40 open source advanced production scheduler
 * platform suite, have look to its licencing options.
 * Web site: http://openi40.org/  
 * Github: https://github.com/openi40/OpenI40Platform
 * We hope you enjoy implementing new amazing projects with it.
 * @author architectures@openi40.org
 *
 */
export class TimeSegment {
  utcStartDateTime?:number=0;
  utcEndDateTime?:number=0;
  effectiveAvailableTime?:number=0;
  public  isValid():boolean {
		return this.utcStartDateTime && this.utcEndDateTime  && this.utcEndDateTime>=this.utcStartDateTime && this.utcStartDateTime>0 && this.utcEndDateTime>0;
	}
  public  isLowerLimited():boolean {
		return this.utcStartDateTime && this.utcStartDateTime>0;
	}

	public   isUpperLimited():boolean {
		return this.utcEndDateTime && this.utcEndDateTime>0;
	}

	public  isLowUpLimited():boolean {
		return this.isLowerLimited() && this.isUpperLimited() && this.isValid();
	}
	public   isInRange(dateTime:number):boolean {
		let inRange:boolean  = false;
		if (this.isValid()) {
			inRange = this.utcStartDateTime<=dateTime  && this.utcEndDateTime>=dateTime;
		} else if (this.isUpperLimited()) {
			inRange = this.utcEndDateTime>=dateTime;
		} else if (this.isLowerLimited()) {
			inRange = this.utcStartDateTime>=dateTime;
		}

		return inRange;
	}

	public isSegmentInRange(range:TimeSegment ):boolean {
		let inRange:boolean  = false;
		if (range.isValid() && this.isValid()) {
			inRange = this.isInRange(range.utcStartDateTime) && this.isInRange(range.utcEndDateTime);
		} else if (range.isLowerLimited() && this.isLowerLimited() && !this.isUpperLimited()) {
			inRange = this.isInRange(range.utcStartDateTime);
		} else if (range.isUpperLimited() && this.isUpperLimited() && !this.isLowerLimited()) {
			inRange = this.isInRange(range.utcEndDateTime);
		}
		return inRange;
	}

	public   isContiguous( range:TimeSegment):boolean {
		let contiguous:boolean = range.isLowerLimited() && this.isUpperLimited() &&  this.utcEndDateTime==range.utcStartDateTime;
		contiguous = contiguous || (this.isLowerLimited() && range.isUpperLimited() && range.utcEndDateTime==this.utcStartDateTime);
		return contiguous;
	}

	public static  IsContiguous<Tr1 extends TimeSegment, Tr2 extends TimeSegment>( range:Tr1,  ranges:Array<Tr2>):boolean  {
		 let rv:boolean = false;
		ranges.forEach((r:Tr2)=>{

				rv = rv || r.isContiguous(range);

		})
		return rv;
	}

	public static   IsOverlappingVector<Tr1 extends TimeSegment, Tr2 extends TimeSegment>( range:Tr1,  ranges:Array<Tr2>):boolean {
    let rv:boolean  = false;
    ranges.forEach((r:Tr2)=>{
      rv = rv || r.IsOverlapping(range);
    })
		return rv;
	}

	public  IsOverlapping( range:TimeSegment):boolean {
		if (this.isContiguous(range)) {
			return false;
		}
		let isOverlapping:boolean  = this.isSegmentInRange(range) || range.isSegmentInRange(this);
		// Particular case of non overlapping => when ranges have Superior and Inferior
		// Bounds equals each other
		isOverlapping = isOverlapping || (range.isLowerLimited() && this.isInRange(range.utcStartDateTime));
		isOverlapping = isOverlapping || (range.isUpperLimited() && this.isInRange(range.utcEndDateTime));
		isOverlapping = isOverlapping || (this.isLowerLimited() && range.isInRange(this.utcStartDateTime));
		isOverlapping = isOverlapping || (this.isUpperLimited() && range.isInRange(this.utcEndDateTime));
		if (isOverlapping) {
			return true;
		}
		if ((this.isLowerLimited() && range.isUpperLimited() && this.utcStartDateTime >= range.utcEndDateTime || (this.isUpperLimited() && range.isLowerLimited() && this.utcEndDateTime === range.utcStartDateTime))) {
			return false;
		}

		if (!isOverlapping && this.isLowUpLimited()) {
			isOverlapping = isOverlapping || (range.isLowerLimited() && this.isInRange(range.utcStartDateTime)) || (range.isUpperLimited() && this.isInRange(range.utcEndDateTime));
		}
		if (!isOverlapping && this.isLowerLimited() && !this.isUpperLimited()) {
			isOverlapping = isOverlapping || range.isInRange(this.utcStartDateTime);

		}
		if (!isOverlapping && !this.isLowerLimited() && this.isUpperLimited()) {
			isOverlapping = isOverlapping || range.isInRange(this.utcEndDateTime);
		}
		if (!isOverlapping && !this.isUpperLimited() && !this.isLowerLimited()) {
			isOverlapping = true;
		}
		return isOverlapping;
	}

	public  equals( range:TimeSegment):boolean {
		let e:boolean = false;
		e = this.isValid() && range.isValid() && this.utcStartDateTime===range.utcStartDateTime && this.utcEndDateTime==range.utcEndDateTime;
		return e;
	}

	public static  Intersection(range:Array<TimeSegment>):TimeSegment {

		let minDateTime:number = null;
    let maxDateTime:number = null;
    if (range) {
      range.forEach((r:TimeSegment)=>{
        if (minDateTime == null) {
          minDateTime = r.utcStartDateTime;
        } else if (r.isLowerLimited()) {
          minDateTime = minDateTime>r.utcStartDateTime ? minDateTime : r.utcStartDateTime;
        }

        if (maxDateTime == null) {
          maxDateTime = r.utcEndDateTime;
        } else if (r.isUpperLimited()) {
          maxDateTime = maxDateTime>r.utcEndDateTime ? maxDateTime : r.utcEndDateTime;
        }
      })
    }


    let _tr:TimeSegment =new TimeSegment();
    _tr.utcStartDateTime=minDateTime;
    _tr.utcEndDateTime=maxDateTime;
		return _tr;
	}





	public  IsSameLimits( tr:TimeSegment):boolean {

		return (this.AreEqual(tr.utcStartDateTime, this.utcStartDateTime) && this.AreEqual(tr.utcEndDateTime, this.utcEndDateTime));
	}

	private AreEqual(startDateTime1:number, startDateTime2:number):boolean  {
		if (startDateTime1 == null && startDateTime2 == null) {
			return true;
		}
		return startDateTime1 != null && startDateTime2 != null ? startDateTime1===startDateTime2 : false;
	}

	public calculateTimeElapsedInMinutes():number {
		let duration:number = -1;
		if (this.isValid()) {
			duration = (this.utcEndDateTime-this.utcStartDateTime)/(1000.0*60.0);

		}
		return duration;
	}
}

File: openi40.scheduler/Model/src/main/java/com/openi40/scheduler/model/time/TimeSegment.java
package com.openi40.scheduler.model.time;

import java.time.temporal.ChronoUnit;
import java.util.Date;
import java.util.List;

import com.openi40.scheduler.common.aps.IApsObject;
import com.openi40.scheduler.common.utils.DateUtil;
import com.openi40.scheduler.model.AbstractApsObject;
/**
 * 
 * This code is part of the OpenI40 open source advanced production scheduler
 * platform suite, have look to its licencing options.
 * Web site: http://openi40.org/  
 * Github: https://github.com/openi40/OpenI40Platform
 * We hope you enjoy implementing new amazing projects with it.
 * @author architectures@openi40.org
 *
 */
public class TimeSegment extends AbstractApsObject implements Cloneable, ITimeSegment {

	private TimeSegmentType Type = TimeSegmentType.values()[0];

	public final TimeSegmentType getType() {
		return Type;
	}

	private IApsObject OwnerObject = null;

	public final IApsObject getOwnerObject() {
		return OwnerObject;
	}

	public final void setOwnerObject(IApsObject value) {
		OwnerObject = value;
	}

	public TimeSegment(TimeSegment tr) {
		super(null);
		setStartDateTime(tr.getStartDateTime());
		setEndDateTime(tr.getEndDateTime());
		Type = tr.getType();
		this.setOwnerObject(tr.getOwnerObject());

	}

	public TimeSegment(TimeSegment tr, TimeSegmentType type, IApsObject ownerObject) {
		super(null);
		setStartDateTime(tr.getStartDateTime());
		setEndDateTime(tr.getEndDateTime());
		Type = type;
		this.setOwnerObject(ownerObject);

	}

	public TimeSegment(TimeSegmentType type, IApsObject ownerObject) {
		super(null);
		Type = type;
		setOwnerObject(ownerObject);
	}

	protected Date StartDateTime = null;

	public Date getStartDateTime() {
		return StartDateTime;
	}

	public void setStartDateTime(Date value) {
		StartDateTime = value;
	}

	protected Date EndDateTime = null;

	public Date getEndDateTime() {
		return EndDateTime;
	}

	public void setEndDateTime(Date value) {
		EndDateTime = value;
	}

	public boolean isValid() {
		return getStartDateTime() != null && getEndDateTime() != null
				&& getEndDateTime().compareTo(getStartDateTime()) >= 0;
	}

	public boolean isInRange(Date dateTime) {
		boolean inRange = false;
		if (isValid()) {
			inRange = getStartDateTime().compareTo(dateTime) <= 0 && getEndDateTime().compareTo(dateTime) >= 0;
		} else if (isUpperLimited()) {
			inRange = getEndDateTime().compareTo(dateTime) >= 0;
		} else if (isLowerLimited()) {
			inRange = getStartDateTime().compareTo(dateTime) <= 0;
		}

		return inRange;
	}

	public boolean isInRange(ITimeSegment range) {
		boolean inRange = false;
		if (range.isValid() && isValid()) {
			inRange = isInRange(range.getStartDateTime()) && isInRange(range.getEndDateTime());
		} else if (range.isLowerLimited() && isLowerLimited() && !isUpperLimited()) {
			inRange = isInRange(range.getStartDateTime());
		} else if (range.isUpperLimited() && isUpperLimited() && !isLowerLimited()) {
			inRange = isInRange(range.getEndDateTime());
		}

		return inRange;
	}

	public boolean isContiguous(ITimeSegment range) {
		boolean contiguous = range.isLowerLimited() && isUpperLimited()
				&& getEndDateTime().compareTo(range.getStartDateTime()) == 0;
		contiguous = contiguous || (isLowerLimited() && range.isUpperLimited()
				&& range.getEndDateTime().compareTo(getStartDateTime()) == 0);
		return contiguous;
	}

	public static <Tr1 extends TimeSegment, Tr2 extends TimeSegment> boolean IsContiguous(Tr1 range, List<Tr2> ranges) {
		boolean rv = false;
		for (Tr2 r : ranges) {
			if (r != range) {
				rv = rv || r.isContiguous(range);
			}
		}
		return rv;
	}

	public static <Tr1 extends TimeSegment, Tr2 extends TimeSegment> boolean IsOverlapping(Tr1 range,
			List<Tr2> ranges) {
		boolean rv = false;
		for (Tr2 r : ranges) {
			if (r != range) {
				rv = rv || r.IsOverlapping(range);
			}
		}
		return rv;
	}

	public boolean IsOverlapping(ITimeSegment range) {
		if (isContiguous(range)) {
			return false;
		}
		boolean isOverlapping = isInRange(range) || range.isInRange(this);
		// Particular case of non overlapping => when ranges have Superior and Inferior
		// Bounds equals each other
		isOverlapping = isOverlapping || (range.isLowerLimited() && isInRange(range.getStartDateTime()));
		isOverlapping = isOverlapping || (range.isUpperLimited() && isInRange(range.getEndDateTime()));
		isOverlapping = isOverlapping || (isLowerLimited() && range.isInRange(getStartDateTime()));
		isOverlapping = isOverlapping || (isUpperLimited() && range.isInRange(getEndDateTime()));
		if (isOverlapping) {
			return true;
		}
		if ((isLowerLimited() && range.isUpperLimited() && getStartDateTime().compareTo(range.getEndDateTime()) >= 0)
				|| (isUpperLimited() && range.isLowerLimited() && getEndDateTime().equals(range.getStartDateTime()))) {
			return false;
		}

		if (!isOverlapping && isLowUpLimited()) {
			isOverlapping = isOverlapping || (range.isLowerLimited() && isInRange(range.getStartDateTime()))
					|| (range.isUpperLimited() && isInRange(range.getEndDateTime()));
		}
		if (!isOverlapping && isLowerLimited() && !isUpperLimited()) {
			isOverlapping = isOverlapping || range.isInRange(getStartDateTime());

		}
		if (!isOverlapping && !isLowerLimited() && isUpperLimited()) {
			isOverlapping = isOverlapping || range.isInRange(getEndDateTime());
		}
		if (!isOverlapping && !isUpperLimited() && !isLowerLimited()) {
			isOverlapping = true;
		}
		return isOverlapping;
	}

	public boolean equals(ITimeSegment range) {
		boolean e = false;
		e = isValid() && range.isValid() && getStartDateTime().equals(range.getStartDateTime())
				&& getEndDateTime().equals(range.getEndDateTime());
		return e;
	}

	public static TimeSegment Intersection(List<TimeSegment> range, TimeSegmentType resultType,
			IApsObject objectOwner) {
		TimeSegment tr = new TimeSegment(resultType, objectOwner);
		Date minDateTime = null;
		Date maxDateTime = null;
		for (TimeSegment r : range) {
			if (minDateTime == null) {
				minDateTime = r.getStartDateTime();
			} else if (r.isLowerLimited()) {
				minDateTime = minDateTime.compareTo(r.getStartDateTime()) > 0 ? minDateTime : r.getStartDateTime();
			}

			if (maxDateTime == null) {
				maxDateTime = r.getEndDateTime();
			} else if (r.isUpperLimited()) {
				maxDateTime = maxDateTime.compareTo(r.getEndDateTime()) > 0 ? maxDateTime : r.getEndDateTime();
			}
		}

		tr.setStartDateTime(minDateTime);
		tr.setEndDateTime(maxDateTime);
		return tr;
	}

	public Object clone() {
		TimeSegment tr = new TimeSegment(this, getType(), getOwnerObject());
		tr.id = id;
		tr.setOwnerObject(getOwnerObject());
		return tr;
	}

	public final boolean isLowerLimited() {
		return getStartDateTime() != null;
	}

	public final boolean isUpperLimited() {
		return getEndDateTime() != null;
	}

	public final boolean isLowUpLimited() {
		return isLowerLimited() && isUpperLimited() && isValid();
	}

	public final boolean IsSameLimits(ITimeSegment tr) {

		return (AreEqual(tr.getStartDateTime(), getStartDateTime()) && AreEqual(tr.getEndDateTime(), getEndDateTime()));
	}

	private boolean AreEqual(Date startDateTime1, Date startDateTime2) {
		if (startDateTime1 == null && startDateTime2 == null) {
			return true;
		}
		return startDateTime1 != null && startDateTime2 != null ? startDateTime1.equals(startDateTime2) : false;
	}

	public double getDurationMinutes() {
		double duration = -1;
		if (isValid()) {
			duration = getStartDateTime().toInstant().until(getEndDateTime().toInstant(), ChronoUnit.MINUTES);

		}
		return duration;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj instanceof TimeSegment) {
			TimeSegment comparison = (TimeSegment) obj;
			boolean _eq = comparison.IsSameLimits(this);
			_eq = _eq && comparison.Type == this.Type;
			_eq = _eq && (code == null && comparison.code == null
					|| (code != null && comparison.code != null && code.equals(comparison.code)));
			_eq = _eq && (description == null && comparison.description == null || (description != null
					&& comparison.description != null && description.equals(comparison.description)));
			_eq = _eq && OwnerObject == comparison.OwnerObject;
			return _eq;
		}
		return super.equals(obj);
	}

	@Override
	public String toString() {
		String v = "TimeSegment{[";
		v += DateUtil.toString(getStartDateTime());
		v += "<-->";
		v += DateUtil.toString(getEndDateTime());
		v += "]";
		v += "Type=" + getType() + "}";

		return v;
	}
}